File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int helper(vector<int>& nums, int n, vector<int>& dp) {
        if (n >= nums.size() - 1) return 0;
        
        if (dp[n] != -1) return dp[n];
        
        int ans = INT_MAX - 1;
        for (int i = 1; i <= nums[n]; i++) {
            int temp = 1 + helper(nums, n + i, dp);
            ans = min(ans, temp);
        }
        
        return dp[n] = ans;
    }
    
    int jump(vector<int>& nums) {
        //Memoization
        vector<int> dp(nums.size(), -1);
        return helper(nums, 0, dp);
        
        // Greedy
//         int currEnd = 0, currFarthest = 0, steps = 0;
//         for (int i = 0; i < nums.size() - 1; i++) {
//             currFarthest = max(currFarthest, i + nums[i]);
            
//             if (i == currEnd) {
//                 steps++;
//                 currEnd = currFarthest;
//             }
//         }
        
//         return steps;
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    int dp[];
    public int jump(int[] nums) {
        
        dp = new int[nums.length];
        for(int i=0;i<nums.length;i++) dp[i] = -1;
        
        return solve(nums, 0);
        
    }
    
    
    public int solve(int[] nums, int currIndex){
        
        if(currIndex >= nums.length - 1) return 0;
        
        if(dp[currIndex] != -1) return dp[currIndex];
        
        int ans = Integer.MAX_VALUE - 1;
        for(int i = 1; i<=nums[currIndex]; i++){
            int jumps = 1 + solve(nums, currIndex + i);
            ans = Math.min(ans, jumps);
        }
        
        return dp[currIndex] = ans;
        
    }
}